src/main.rs

```diff
@@ -129,116 +130,104 @@ fn metadata(path: &Path) -> Result<Metadata, Error> {
-            // TODO: consider parsing units like GB, KB ...
-            let size = match matches
-                .value_of("maxsize")
-                .and_then(|s| s.parse::<u64>().ok())
-            {
-                Some(s) => s * 1024 * 1024,
-                None => {
-                    anyhow::bail!("maxsize has to be a number");
+                    )
                 }
+                Ok(cleaned_amount) => info!(
+                    "Cleaned {} from {project_path:?}",
+                    format_bytes(cleaned_amount)
+                ),
+                Err(e) => error!(
+                    "{:?}",
+                    e.context(format!("Failed to clean {project_path:?}"))
+                ),
             };
-
-            for project_path in &paths {
-                match remove_older_until_fits(project_path, size, dry_run) {
-                    Ok(cleaned_amount) if dry_run => {
-                        info!(
-                            "Would clean: {} from {project_path:?}",
-                            format_bytes(cleaned_amount)
-                        )
-                    }
-                    Ok(cleaned_amount) => info!(
-                        "Cleaned {} from {project_path:?}",
+        }
+    } else if let Criterion::MaxSize(size) = criterion {
+        for project_path in &paths {
+            match remove_older_until_fits(project_path, size, dry_run) {
+                Ok(cleaned_amount) if dry_run => {
+                    info!(
+                        "Would clean: {} from {project_path:?}",
                         format_bytes(cleaned_amount)
-                    ),
-                    Err(e) => error!("Failed to clean {:?}: {:?}", project_path, e),
-                };
-            }
-        } else {
-            let keep_duration = if matches.is_present("file") {
-                let ts = Timestamp::load(path.as_path()).expect("Failed to load timestamp file");
-                Duration::from(ts)
-            } else {
-                let days_to_keep: u64 = matches
-                    .value_of("time")
-                    .expect("--time argument missing")
-                    .parse()
-                    .expect("Invalid time format");
-                Duration::from_secs(days_to_keep * 24 * 3600)
+                    )
+                }
+                Ok(cleaned_amount) => info!(
+                    "Cleaned {} from {project_path:?}",
+                    format_bytes(cleaned_amount)
+                ),
+                Err(e) => error!("Failed to clean {:?}: {:?}", project_path, e),
             };
+        }
+    } else {
+        let keep_duration = if let Criterion::File = criterion {
+            let ts = Timestamp::load(path.as_path()).expect("Failed to load timestamp file");
+            Duration::from(ts)
+        } else if let Criterion::Time(days_to_keep) = criterion {
+            Duration::from_secs(days_to_keep * 24 * 3600)
+        } else {
+            unreachable!();
```

Comment:
Is it possible to get rid of this `unreachable` by using `match` instead of `if let`? That would let us get a compile error if we add a new Criterion without handling it here.

Reply:
I was also unsatisfied with this piece of code (and some others xD).

```rust
match criterion {
    Criterion::Time(...) => ...,
    Criterion::File => ...,
    _ => unreachable!(),
}
```

Are you suggesting I replace the `_` wildcard by the combination of the rest of the variants?

If I use `_` I won't type check.

Reply:
Yes, the rest of the variants - I don't mean `Criterion::Stamp => unreachable!()`, but instead migrating the whole series of `if-else` blocks to a match. You can make the match arms separate helper functions if that helps reduce nesting.

Reply:
actually on second thought this PR is large enough - `unreachable` is fine for now :) no need to change anything